packages/shared/src/components/HorizontalScroll/useCalculateVisibleElements.ts

```diff
@@ -0,0 +1,72 @@
+import { MutableRefObject, useCallback, useEffect, useState } from 'react';
+import { useToggle } from '../../hooks/useToggle';
+import { useEventListener } from '../../hooks';
+
+interface UseCalculateVisibleElementsProps<El extends HTMLElement> {
+  ref: MutableRefObject<El | null>;
+}
+
+interface UseCalculateVisibleElements {
+  calculateVisibleCards: () => void;
+  isOverflowing: boolean;
+  elementsCount: number;
+  scrollableElementWidth: number;
+}
+
+export const useCalculateVisibleElements = <El extends HTMLElement>({
+  ref,
+}: UseCalculateVisibleElementsProps<El>): UseCalculateVisibleElements => {
+  const [isOverflowing, setIsOverflowing] = useToggle(false);
+  const [elementsCount, setElementsCount] = useState<number>(0);
+  const [scrollableElementWidth, setScrollableElementWidth] =
+    useState<number>(0);
+
+  const calculateVisibleCards = useCallback(() => {
+    const currentRef = ref.current;
+    if (
+      !currentRef ||
+      !currentRef.firstElementChild ||
+      !(currentRef.firstElementChild instanceof HTMLElement)
+    ) {
+      return;
+    }
+
+    setIsOverflowing(currentRef.scrollWidth > currentRef.clientWidth);
+
+    if (currentRef && currentRef.firstElementChild) {
+      const element = currentRef.firstElementChild;
+      const elementWidth = element.offsetWidth;
+      const elementMarginRight = parseFloat(
+        getComputedStyle(element).marginRight || '0',
+      );
+      const elementMarginLeft = parseFloat(
+        getComputedStyle(element).marginLeft || '0',
+      );
+      const containerGap = parseFloat(getComputedStyle(currentRef).gap || '0');
+      const elementWidthWithMarginsAndGap =
+        elementWidth + elementMarginRight + elementMarginLeft + containerGap;
+
+      setScrollableElementWidth(elementWidthWithMarginsAndGap);
+
+      const mainContainerWidth = currentRef.offsetWidth;
+      setElementsCount(
+        Math.floor(mainContainerWidth / elementWidthWithMarginsAndGap),
+      );
+    }
+  }, [ref, setIsOverflowing]);
```

Comment:
The ref won't trigger anymore on client side right, making the useEffect actually not re-trigger as well? (didn't probe any issues locally?)

Reply:
Yeah, this won't trigger since it is a ref. Just the linter asking it to be part of the deps since it came from the props.

Reply:
Ah yes, but my concern is wider in that case.
Since that won't ever trigger does the calculation ever happen in the ref is not yet set for instance but you rely on it to give accurate adjustments?

Reply:
Yeah just saying if it's not available on server render yet, it will never retrigger it.
But if you tested it and had no issues it's probably fine.

Reply:
Yeah, the calculation result will be useful once the user can click the left/right buttons. Server-side we should be good.